app/jobs/appointment_notification/worker.rb

```diff
@@ -67,20 +67,6 @@ def send_message(notification, communication_type)
       raise UnknownCommunicationType, "#{self.class.name} is not configured to handle communication type #{communication_type}"
     end
 
-    log_info = {
-      class: self.class.name,
-      msg: "send_message",
-      failed: !!notification_service.failed?,
-      error: notification_service.error,
-      sender: medication_reminder_sms_sender || "default",
-      communication_type: communication_type,
-      notification_id: notification.id
-    }
-
-    logger.info log_info
-
-    return if notification_service.failed?
```

Comment:
The big change in behavior is that notification failure now raises an exception. This will trigger a retry automatically as long as the notification service is being invoked in Sidekiq. Are there adverse consequences to this? What kinds of errors can we expect to see? Will we be retrying notifications in situations where it will never succeed? Eg. sending to an invalid phone number?

Does Twilio's exceptions give us some granular information (like discrete error codes) that we can use to selectively decide _when_ we want to retry (eg. network blip), and when we don't (eg. bad phone number)?

Reply:
I believe we only expect twilio errors on network errors, but that's largely guesswork. This code was all in place before I started working on it. When I found it, this worker was rescuing from the twilio error and logging. IMO, this class shouldn't need to know anything about the twilio error. That should all be handled from the twilio client. The twilio client is used in other places, so we had multiple `rescue TwilioError` in our code, which strikes me as a bad pattern. And it will only get worse when we add Imo, which is why I'm making changes. I think the other option is to have a `NotificationService` that handles the Twilio and Imo clients. We can pass some context into it and pass it into the error class. I don't think we should have this logging here, though. It would be better to have it all in one place (i.e., sentry log for the error) not spread out across Sentry and rails logs. I'll look into what would be causing twilio errors (i.e., bad number or just network error) and try to come up with a better solution.

Reply:
The major point I wanna think through is - will this make our code retry jobs again and again that are never destined to pass. If we can find out by taking a look at the Twilio gem or API error codes, etc, it may help us understand better, and my concerns would be addressed 👍

Reply:
I agree with that point, and I'll try to figure that out. We did have a bug that resent the same messages every day because they never succeeded, and I don't want to recreate the problem. My point is really more about keeping the error handling as close to the origin as possible and not spreading its logging across systems.

Reply:
As best I can tell, Sentry only keeps logs for 90 days. Notifications have been turned off for some of that time, but I only see two instances of twilio errors. Both are for a bad number. I can try to avoid retries for this case, but it should also capture network errors. 
https://sentry.io/organizations/resolve-to-save-lives/issues/2443427939/?end=2021-07-02T10%3A51%3A54&environment=production&project=1793711&query=Error+while+processing+notification&start=2021-04-03T10%3A51%3A54&utc=true